### ISSUE TITLE
Crash after downloading files

### ISSUE THREAD

**Issue by johneriksson** [casperjs/casperjs#1255]
Not sure if I'm supposed to post here or on PhantomJS but I have a CasperJS-program that reads a .txt file with about 40 URLs. The program then opens all of these URLs and downloads the file there (can be jpg, pdf or bmp). But every time the program runs, it crashes after downloading the files. In some cases it downloads all before crashing, other times 2-3 are missing. But the program will always crash.

Thinking it was a memory-issue, I tried closing or killing the opened page after it was done downloading but that just seemed to kill the entire casper program.

Info that might help:
OS - Windows 8.1
PhantomJS and CasperJS official binaries

My program (shortened):

```
var fs = require('fs');
var data = fs.read('links.txt');
var links = data.split(",");

var casper = require('casper').create({
    pageSettings: {
        webSecurityEnabled: false
    }
});

casper.start().each(links, function (self, link) {
    self.thenOpen(link, function () {
        // url is found in an <a>-tag

        this.echo("Downloading: " + url);
        var urlslices = url.split("/");
        var filename = "images/" + urlslices[urlslices.length - 1];
        this.download(url, filename);
    });
});

casper.run();
```

The .dmp file:

```
Dump Summary
------------
Dump File:  e0a550f9-b720-4f31-9c41-d31bee12bb01.dmp : C:\Users\John\AppData\Local\Temp\e0a550f9-b720-4f31-9c41-d31bee12bb01.dmp
Last Write Time:    2015-06-22 10:49:24
Process Name:   phantomjs.exe : C:\Windows\phantomjs.exe
Process Architecture:   x86
Exception Code: 0xC0000005
Exception Information:  The thread tried to read from or write to a virtual address for which it does not have the appropriate access.
Heap Information:   Not Present

System Information
------------------
OS Version: 6.3.9600
CLR Version(s): 

Modules
-------
Module Name Module Path Module Version
----------- ----------- --------------
phantomjs.exe   C:\Windows\phantomjs.exe    2.0.0.0
ntdll.dll   C:\Windows\System32\ntdll.dll   6.3.9600.17736
kernel32.dll    C:\Windows\System32\kernel32.dll    6.3.9600.17415
KERNELBASE.dll  C:\Windows\System32\KERNELBASE.dll  6.3.9600.17415
ws2_32.dll  C:\Windows\System32\ws2_32.dll  6.3.9600.17415
winspool.drv    C:\Windows\System32\winspool.drv    6.3.9600.17415
gdi32.dll   C:\Windows\System32\gdi32.dll   6.3.9600.17415
oleaut32.dll    C:\Windows\System32\oleaut32.dll    6.3.9600.17560
imm32.dll   C:\Windows\System32\imm32.dll   6.3.9600.17415
winmm.dll   C:\Windows\System32\winmm.dll   6.3.9600.17415
user32.dll  C:\Windows\System32\user32.dll  6.3.9600.17415
shell32.dll C:\Windows\System32\shell32.dll 6.3.9600.17824
ole32.dll   C:\Windows\System32\ole32.dll   6.3.9600.17415
advapi32.dll    C:\Windows\System32\advapi32.dll    6.3.9600.17415
crypt32.dll C:\Windows\System32\crypt32.dll 6.3.9600.17475
nsi.dll C:\Windows\System32\nsi.dll 6.3.9600.17415
rpcrt4.dll  C:\Windows\System32\rpcrt4.dll  6.3.9600.17415
msvcrt.dll  C:\Windows\System32\msvcrt.dll  7.0.9600.17415
combase.dll C:\Windows\System32\combase.dll 6.3.9600.17415
msctf.dll   C:\Windows\System32\msctf.dll   6.3.9600.17706
WINMMBASE.dll   C:\Windows\System32\WINMMBASE.dll   6.3.9600.17415
shlwapi.dll C:\Windows\System32\shlwapi.dll 6.3.9600.17415
sechost.dll C:\Windows\System32\sechost.dll 6.3.9600.17734
msasn1.dll  C:\Windows\System32\msasn1.dll  6.3.9600.17415
sspicli.dll C:\Windows\System32\sspicli.dll 6.3.9600.17415
cfgmgr32.dll    C:\Windows\System32\cfgmgr32.dll    6.3.9600.17415
devobj.dll  C:\Windows\System32\devobj.dll  6.3.9600.17415
CRYPTBASE.dll   C:\Windows\System32\CRYPTBASE.dll   6.3.9600.17415
bcryptPrimitives.dll    C:\Windows\System32\bcryptPrimitives.dll    6.3.9600.17415
dbghelp.dll C:\Windows\System32\dbghelp.dll 6.3.9600.17787
SHCore.dll  C:\Windows\System32\SHCore.dll  6.3.9600.17666
profapi.dll C:\Windows\System32\profapi.dll 6.3.9600.17415
kernel.appcore.dll  C:\Windows\System32\kernel.appcore.dll  6.3.9600.17415
uxtheme.dll C:\Windows\System32\uxtheme.dll 6.3.9600.17415
tiptsf.dll  C:\Program Files (x86)\Common Files\microsoft shared\Ink\tiptsf.dll 6.3.9600.17415
dwmapi.dll  C:\Windows\System32\dwmapi.dll  6.3.9600.17415
cryptsp.dll C:\Windows\System32\cryptsp.dll 6.3.9600.17415
rsaenh.dll  C:\Windows\System32\rsaenh.dll  6.3.9600.17415
bcrypt.dll  C:\Windows\System32\bcrypt.dll  6.3.9600.17415
IPHLPAPI.DLL    C:\Windows\System32\IPHLPAPI.DLL    6.3.9600.17415
winnsi.dll  C:\Windows\System32\winnsi.dll  6.3.9600.17415
dhcpcsvc.dll    C:\Windows\System32\dhcpcsvc.dll    6.3.9600.17415
dhcpcsvc6.DLL   C:\Windows\System32\dhcpcsvc6.DLL   6.3.9600.17415
wlanapi.dll C:\Windows\System32\wlanapi.dll 6.3.9600.17415
userenv.dll C:\Windows\System32\userenv.dll 6.3.9600.17415
winhttp.dll C:\Windows\System32\winhttp.dll 6.3.9600.17415
mswsock.dll C:\Windows\System32\mswsock.dll 6.3.9600.17415
dnsapi.dll  C:\Windows\System32\dnsapi.dll  6.3.9600.17481
rasadhlp.dll    C:\Windows\System32\rasadhlp.dll    6.3.9600.17415
FWPUCLNT.DLL    C:\Windows\System32\FWPUCLNT.DLL    6.3.9600.17485
```

**Comment by kensoh** [comment]
Hi @johneriksson not sure if this is still happening for you or if the issue is fixed. I'm aware PhantomJS has large memory consumption. Could it work for you if you wrap the iteration around shell or batch script? I do it something like this to process batch automation. So that the PhantomJS/CasperJS process is restarted at each iteration. Just my preference, I think it should be more stable to run batch.

https://github.com/tebelorg/TagUI/blob/master/src/tagui.cmd

**Comment by kensoh** [comment]
Hi @johneriksson not sure if this issue is resolved for you with latest CasperJS v1.1.4 and PhantomJS v2.1.1? Closing this for now but please reopen if it is still happening. Thanks for raising this! For direct downloading of files from URLs maybe using CURL is more resilient solution.
